File: apps/hermes-universal/src/store/find-in-page.ts
import { invoke } from '@tauri-apps/api/core'
import { listen, type UnlistenFn } from '@tauri-apps/api/event'

import { stepOrdinal } from '@/lib/find-in-page'
import { clearDomFindSelection, countDomTextMatches, domFind, domFindSupported } from '@/lib/find-in-page-dom'
import { PLATFORM } from '@/lib/platform'
import { atom } from '@/store/atom'

/**
 * Find-in-page state (MJXHRM-49, MJXHRM-387).
 *
 * Desktop drives Electron's `webContents.findInPage`. Universal has TWO doors
 * onto the same idea and picks between them once, here:
 *
 *   - **Linux** invokes the Rust commands in `src-tauri/src/find_in_page.rs`,
 *     which reach through to WebKitGTK's own find controller. Highlight-all,
 *     engine-accurate counts.
 *   - **Everywhere else** (macOS, Windows, Android) uses `window.find` plus a
 *     text scan — see `lib/find-in-page-dom.ts` for why that is the trade
 *     instead of three native engine bindings.
 *
 * The ordinal is OURS on both paths. WebKitGTK reports how many matches exist
 * and never which one is selected, so "3 of 12" is counted here as the user
 * steps — see `stepOrdinal`.
 */

/** Emitted by the Rust side with the match count for the current query. */
const FOUND_IN_PAGE_EVENT = 'hermes://found-in-page'

export interface FindInPageState {
  active: boolean
  query: string
  matchOrdinal: number
  matchCount: number
}

const EMPTY: FindInPageState = { active: false, matchCount: 0, matchOrdinal: 0, query: '' }

export const $findInPage = atom<FindInPageState>({ ...EMPTY })

/** True when the native (Rust/WebKitGTK) path is the one to drive. */
function usesNativeFind(): boolean {
  return PLATFORM === 'linux'
}

/**
 * Whether this build can search the page at all.
 *
 * Checked here rather than letting a call fail, so ⌘F simply does nothing on an
 * engine that can't honour it instead of flashing a bar that reports zero
 * matches for everything. In practice both branches are true on every target we
 * ship; the check survives for plain-browser dev and for any engine without
 * `window.find`.
 */
export function findInPageSupported(): boolean {
  return usesNativeFind() || domFindSupported()
}

function callFind(query: string, forward: boolean, findNext: boolean): void {
  if (usesNativeFind()) {
    void invoke('find_in_page', { findNext, forward, query }).catch(() => undefined)

    return
  }

  // `fromStart` on a fresh query: otherwise `window.find` resumes from wherever
  // the previous query left the caret and silently skips everything above it.
  domFind(query, { backwards: !forward, fromStart: !findNext })

  // The engine event never fires on this path, so report the count ourselves —
  // asynchronously, to keep the "a late result can't resurrect a cleared query"
  // guard in `updateFindResults` on the same footing as the native path.
  const counted = countDomTextMatches(query)

  queueMicrotask(() => updateFindResults(counted))
}

function callStop(): void {
  if (usesNativeFind()) {
    void invoke('stop_find_in_page').catch(() => undefined)

    return
  }

  clearDomFindSelection()
}

export function openFindBar(): void {
  if (!findInPageSupported()) {
    return
  }

  $findInPage.set({ ...EMPTY, active: true })
}

export function closeFindBar(): void {
  // Already closed: don't re-issue the stop. Escape is a shared gesture (dialogs
  // and the session switcher claim it too), so a stray second close must not
  // reach into the engine again.
  if (!$findInPage.get().active) {
    return
  }

  $findInPage.set({ ...EMPTY })
  // Ends the search AND drops the highlight — a bar that closes over a still-
  // highlighted page has not really closed.
  callStop()
}

export function setFindQuery(query: string): void {
  const prev = $findInPage.get()

  // Never search for a closed bar. The component clears its debounce on close,
  // but a timer that already fired (or any late caller) must not re-issue a find
  // and re-highlight the page after the user pressed Escape.
  if (!prev.active) {
    return
  }

  if (!query) {
    $findInPage.set({ ...prev, matchCount: 0, matchOrdinal: 0, query: '' })
    callStop()

    return
  }

  // Ordinal 0 until the engine answers: a fresh query has selected nothing yet,
  // and carrying the previous query's position over would be a lie for a frame.
  $findInPage.set({ ...prev, matchOrdinal: 0, query })
  callFind(query, true, false)
}

function step(direction: 'backward' | 'forward'): void {
  const state = $findInPage.get()

  if (!state.active || !state.query) {
    return
  }

  $findInPage.set({ ...state, matchOrdinal: stepOrdinal(state.matchOrdinal, state.matchCount, direction) })
  callFind(state.query, direction === 'forward', true)
}

export function findNext(): void {
  step('forward')
}

export function findPrevious(): void {
  step('backward')
}

/**
 * Called when the engine reports a count for the current query.
 *
 * A count arriving for a query the user has already cleared is dropped: the
 * search is async, so a result for the previous keystroke can land after the
 * bar was emptied and would resurrect a stale counter.
 */
export function updateFindResults(count: number): void {
  const prev = $findInPage.get()

  if (!prev.active || !prev.query) {
    return
  }

  const matchCount = Number.isFinite(count) && count > 0 ? Math.floor(count) : 0

  $findInPage.set({
    ...prev,
    matchCount,
    // First result for a fresh query: the engine has selected match one.
    matchOrdinal: matchCount === 0 ? 0 : prev.matchOrdinal || 1
  })
}

// The result subscription is process-wide, not per-mount: the find bar lives in
// the global overlay set and the shell remounts it on a connection re-home while
// route changes keep it alive. Refcount the single listener so a remount cannot
// stack duplicates — every stacked listener would re-dispatch the same result
// and, worse, outlive its component.
let listenerRefs = 0
let detachListener: (() => void) | undefined
let pendingDetach: Promise<UnlistenFn> | undefined

/**
 * Subscribe to match counts. Returns a release fn; the underlying listener is
 * installed on the first subscriber and removed when the last one releases. Safe
 * to call from an effect with a `[]` dep list.
 */
export function initFindInPageListener(): () => void {
  listenerRefs += 1

  if (listenerRefs === 1) {
    // `listen` resolves asynchronously, so a mount/unmount inside one tick would
    // otherwise leave the listener attached with the refcount already at zero.
    pendingDetach = listen<number>(FOUND_IN_PAGE_EVENT, event => updateFindResults(event.payload))

    void pendingDetach
      .then(stop => {
        if (listenerRefs > 0) {
          detachListener = stop
        } else {
          stop()
        }
      })
      .catch(() => undefined)
  }

  let released = false

  return () => {
    // Guard double-release: React can invoke a cleanup once, but a caller
    // holding the fn shouldn't be able to drive the refcount negative.
    if (released) {
      return
    }

    released = true
    listenerRefs -= 1

    if (listenerRefs === 0) {
      detachListener?.()
      detachListener = undefined
      pendingDetach = undefined
    }
  }
}

/** Test seam: number of live subscriptions (0 or 1 in practice). */
export function findInPageListenerCount(): number {
  return listenerRefs
}

File: apps/hermes-universal/src-tauri/src/find_in_page.rs
//! Find-in-page (MJXHRM-49) — the browser's own incremental search over the
//! rendered page.
//!
//! Desktop drives Electron's `webContents.findInPage`. Tauri has no equivalent,
//! so this reaches through `WebviewWindow::with_webview` to the underlying
//! WebKitGTK `WebView` and drives its `WebKitFindController` directly — the same
//! engine-level machinery Electron's API is a wrapper around.
//!
//! **Why not a DOM overlay.** The obvious alternative is to walk the document in
//! JavaScript and wrap matches in `<mark>`s. It cannot work here: the chat
//! transcript, the file tree and the editor are all virtualized, so most of the
//! text the user is searching for is not in the DOM at the moment they search
//! for it. A DOM scan would confidently report "no matches" for text that is
//! plainly in the conversation. The engine's own find has the same blind spot
//! for unmounted rows, but it is at least the same search the platform performs
//! everywhere else, and it highlights and scrolls without us reimplementing
//! either.
//!
//! **This module is the LINUX path, not the whole feature.** `with_webview`
//! hands back a platform-specific handle: `WKWebView` on macOS (no public find
//! API at all — `NSTextFinder` does not apply to web content), `ICoreWebView2`
//! on Windows (its `Find` API landed only in recent runtime versions and needs a
//! direct `webview2-com` dependency), and a `WebView` object on Android (JNI).
//! Three separate native bindings, none of which can even be COMPILED on the
//! Linux host this is developed on — so every other target is served instead by
//! a portable `window.find` path in `src/lib/find-in-page-dom.ts`, and the
//! frontend picks between the two in `src/store/find-in-page.ts`. Native
//! bindings for the remaining engines stay open as MJXHRM-302; they would buy
//! highlight-all and exact counts, not the feature itself.
//!
//! **One ordinal caveat.** WebKitGTK reports the match COUNT (`found-text`,
//! `counted-matches`) but never which match is currently selected — there is no
//! equivalent of Electron's `activeMatchOrdinal`. The frontend therefore counts
//! the steps itself (`store/find-in-page.ts`); this module reports only what the
//! engine actually knows.

/// Emitted to the searching window with the match count for the current query.
pub const FOUND_IN_PAGE_EVENT: &str = "hermes://found-in-page";

#[cfg(target_os = "linux")]
mod linux {
    use std::collections::HashSet;
    use std::sync::{Mutex, OnceLock};

    use tauri::Emitter;

    use super::FOUND_IN_PAGE_EVENT;

    /// Windows whose find controller already has our signal handlers.
    ///
    /// The handlers are connected inside `with_webview`, which runs per call —
    /// without this, every ⌘F would stack another listener on the same
    /// controller and each result would be emitted N times.
    fn wired() -> &'static Mutex<HashSet<String>> {
        static WIRED: OnceLock<Mutex<HashSet<String>>> = OnceLock::new();
        WIRED.get_or_init(|| Mutex::new(HashSet::new()))
    }

    /// Case-insensitive and wrapping — what every find bar does, and what the
    /// user means by "find" without being asked.
    fn find_options() -> u32 {
        (webkit2gtk::FindOptions::CASE_INSENSITIVE | webkit2gtk::FindOptions::WRAP_AROUND).bits()
    }

    /// No practical cap: the counter should say how many matches there ARE, and
    /// a truncated total is worse than a slightly slower count.
    const MAX_MATCHES: u32 = u32::MAX;

    /// What to do to the controller, once we are on the main thread with it.
    pub enum Op {
        Next,
        Previous,
        Search(String),
        Stop,
    }

    pub fn run(window: tauri::WebviewWindow, op: Op) -> Result<(), String> {
        let label = window.label().to_string();
        let emitter = window.clone();

        window
            .with_webview(move |platform_webview| {
                use webkit2gtk::{FindControllerExt, WebViewExt};

                let Some(controller) = platform_webview.inner().find_controller() else {
                    return;
                };

                // First search in this window: subscribe to the results. Both
                // signals matter — `failed-to-find-text` is the ONLY way a query
                // with no matches is reported, so without it the counter would
                // keep showing the previous query's total.
                let first_time = wired()
                    .lock()
                    .map(|mut set| set.insert(label.clone()))
                    .unwrap_or(false);

                if first_time {
                    let found = emitter.clone();
                    controller.connect_found_text(move |_, match_count| {
                        let _ = found.emit(FOUND_IN_PAGE_EVENT, match_count);
                    });

                    let counted = emitter.clone();
                    controller.connect_counted_matches(move |_, match_count| {
                        let _ = counted.emit(FOUND_IN_PAGE_EVENT, match_count);
                    });

                    let failed = emitter.clone();
                    controller.connect_failed_to_find_text(move |_| {
                        let _ = failed.emit(FOUND_IN_PAGE_EVENT, 0u32);
                    });
                }

                match &op {
                    Op::Next => controller.search_next(),
                    Op::Previous => controller.search_previous(),
                    Op::Search(query) => {
                        // `search` highlights and jumps; `count_matches` is what
                        // fills in the total, and it is a separate call because
                        // `found-text` reports the count of the CURRENT batch
                        // rather than the document total.
                        controller.search(query, find_options(), MAX_MATCHES);
                        controller.count_matches(query, find_options(), MAX_MATCHES);
                    }
                    // Ends the search AND drops the highlight/selection, which is
                    // what Escape has to mean — a bar that closes over a still-
                    // highlighted page has not really closed.
                    Op::Stop => controller.search_finish(),
                }
            })
            .map_err(|e| format!("find-in-page unavailable: {e}"))
    }
}

/// Search the window's rendered page.
///
/// `find_next` distinguishes "step to the next match of the query I already
/// gave you" from "search for this query from scratch" — the frontend passes
/// false on a fresh query and true on Enter / ⌘G, exactly as the Electron
/// bridge does.
#[cfg(target_os = "linux")]
#[tauri::command]
pub async fn find_in_page(
    window: tauri::WebviewWindow,
    query: String,
    forward: Option<bool>,
    find_next: Option<bool>,
) -> Result<(), String> {
    let op = if find_next.unwrap_or(false) {
        if forward.unwrap_or(true) {
            linux::Op::Next
        } else {
            linux::Op::Previous
        }
    } else {
        linux::Op::Search(query)
    };

    linux::run(window, op)
}

/// Stop searching and clear the highlight.
#[cfg(target_os = "linux")]
#[tauri::command]
pub async fn stop_find_in_page(window: tauri::WebviewWindow) -> Result<(), String> {
    linux::run(window, linux::Op::Stop)
}

// Every other platform needs its own engine binding (see the module docs) and
// takes the portable `window.find` path instead, so these commands are never
// invoked there — `store/find-in-page.ts` branches on `PLATFORM === 'linux'`
// BEFORE calling, and does not read this string. They stay registered anyway so
// a stray call returns a clear, catchable error rather than an "unknown
// command" that looks like a build fault.
#[cfg(not(target_os = "linux"))]
#[tauri::command]
pub async fn find_in_page(
    _window: tauri::WebviewWindow,
    _query: String,
    _forward: Option<bool>,
    _find_next: Option<bool>,
) -> Result<(), String> {
    Err("unsupported_platform".to_string())
}

#[cfg(not(target_os = "linux"))]
#[tauri::command]
pub async fn stop_find_in_page(_window: tauri::WebviewWindow) -> Result<(), String> {
    Err("unsupported_platform".to_string())
}
